kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerExtension.java

```diff
@@ -0,0 +1,453 @@
+                    } catch (WakeupException ex) {
+                        logger.trace("Kafka wait interrupted");
+                    } catch (Exception ex) {
+                        logger.error("Error polling kafka consumer", ex);
+                    }
+                } finally {
+                    consumerLock.unlock();
+                }
+                if (consumerReady.get() && !events.isEmpty()) {
+                    if (logger.isDebugEnabled()) {
+                        Map<String, Integer> eventsPerTopic = new HashMap<>();
+                        for (ConsumerRecord<String, byte[]> event : events) {
+                            eventsPerTopic.compute(event.topic(), (k, v) -> v == null ? 1 : v++);
+                        }
+                        logger.debug("Number of events received per topic {}", eventsPerTopic);
+                    }
+                    for (ConsumerRecord<String, byte[]> event : events) {
+                        notifyService.submit(() -> processEvent(event));
+                    }
+                }
+            }
+        }
+        logger.trace("Kafka polling stopped");
+    }
+
+    private boolean isSubscribed() {
+        changeRegistrationLock.lock();
+        try {
+            return !topic2Message.isEmpty() || !topic2Signal.isEmpty();
+        } finally {
+            changeRegistrationLock.unlock();
+        }
+
+    }
+
+    @FunctionalInterface
+    private static interface Signaller {
+        void signalEvent(String deploymentId, String signalName, Object data);
+    }
+
+    private void processEvent(ConsumerRecord<String, byte[]> event) {
+        changeRegistrationLock.lock();
+        try {
+            processEvent(topic2Signal, event, this::signalEvent);
+            processEvent(topic2Message, event,
+                    (deployment, signalName, data) -> signalEvent(deployment, "Message-" + signalName, data));
+        } finally {
+            changeRegistrationLock.unlock();
+        }
+    }
+    
+    private void signalEvent  (String deployment, String signalName, Object data) {
+        processService.signalEvent(deployment,signalName, data);
+    }
+
+    private <T extends SignalDescBase> void processEvent(Map<String, Map<T, Collection<String>>> topic2SignalBase,
+                                                         ConsumerRecord<String, byte[]> event,
+                                                         Signaller signaller) {
+        Map<T, Collection<String>> signalInfo = topic2SignalBase.get(event.topic());
+        if (signalInfo != null) {
+            for (Map.Entry<T, Collection<String>> entry : signalInfo.entrySet())
+                try {
+                    String signalName = entry.getKey().getName();
+                    for (String deploymentId : entry.getValue()) {
+                        CloudEvent<?> cloudEvent = CloudEvent.read(event.value(), getDataClass(deploymentId, entry
+                                .getKey()));
```

Comment:
this is ok but has a limitation. we won't support in kafka events typed json unmarshaller. 
```
{
    map : {
         "key1" : {
               "com.my.simple.type" : {
                       "field" : "hello"
               }
          }
    }

```
(for instance when we have Map, List so the values are not clear and require types being in the json. 
This limitation should be in the BAPL.

Reply:
Limitation docummented in bapl